pkg/search/controller.go

```diff
@@ -161,46 +163,210 @@ func (c *Controller) handleErr(err error, key interface{}) {
-	// STEP0:  stop informer manager for the cluster which is not referenced by any `SearchRegistry` object.
-	v, ok := c.clusterRegistry.Load(cluster)
-	if !ok {
-		klog.Infof("Cluster %s is not registered", cluster)
-		return nil
+func (c *Controller) getClusterMatchedRegistries(cluster *clusterv1alpha1.Cluster) (indexedByName, matched map[string]*searchv1alpha1.ResourceRegistry, err error) {
+	client := c.informerFactory.Search().V1alpha1().ResourceRegistries().Lister()
+	var registries []*searchv1alpha1.ResourceRegistry
+	if registries, err = client.List(labels.Everything()); err != nil {
+		klog.Errorf("List resource registries for reconciling cluster failed, err: %s", err)
+		return
+	}
+	if len(registries) == 0 {
+		klog.Infof("No resource registries, no need to reconcile cluster")
+		return
 	}
+	indexedByName = make(map[string]*searchv1alpha1.ResourceRegistry, len(registries))
+	matched = make(map[string]*searchv1alpha1.ResourceRegistry, len(registries))
+	for i, registry := range registries {
+		indexedByName[registry.Name] = registries[i]
+		if util.ClusterMatches(cluster, registry.Spec.TargetCluster) {
+			matched[registry.Name] = registries[i]
+		}
+	}
+	return
+}
 
-	cr := v.(clusterRegistry)
-	if cr.unregistry() {
-		klog.Infof("try to stop cluster informer %s", cluster)
-		c.InformerManager.Stop(cluster)
-		return nil
+func (c *Controller) getRegistryAddedResources(registry *searchv1alpha1.ResourceRegistry, cr *clusterRegistry, added []schema.GroupVersionResource) []schema.GroupVersionResource {
+	name := registry.Name
+	resourcesToWatch := c.getResources(registry.Spec.ResourceSelectors)
+	for _, resource := range resourcesToWatch {
+		if resourcesOnWatching, exist := cr.resources[resource]; exist {
+			resourcesOnWatching[name] = struct{}{}
+		} else {
+			added = append(added, resource)
+			cr.resources[resource] = registrySet{name: struct{}{}}
+		}
+	}
+	return added
+}
+
+func (c *Controller) getClusterRegistriesModification(registries map[string]*searchv1alpha1.ResourceRegistry, matched map[string]*searchv1alpha1.ResourceRegistry, cr *clusterRegistry, added []string, removed []string) (updated *clusterRegistry, addedResources []schema.GroupVersionResource, removedResources []schema.GroupVersionResource) {
+	defer func() {
+		updated = cr
+	}()
+	if cr == nil {
+		cr = &clusterRegistry{
+			resources:  make(map[schema.GroupVersionResource]registrySet),
+			registries: make(registrySet)}
+	}
+	for _, name := range added {
+		cr.registries[name] = struct{}{}
+		addedResources = c.getRegistryAddedResources(registries[name], cr, addedResources)
+	}
+	for _, name := range removed {
+		delete(cr.registries, name)
+		for resource, previousRegistries := range cr.resources {
+			delete(previousRegistries, name)
+			if len(previousRegistries) == 0 {
+				removedResources = append(removedResources, resource)
+			}
+		}
+		for _, resource := range removedResources {
+			delete(cr.resources, resource)
+		}
+	}
```

Comment:
Do we need to handle deletions first before handling additions, because the added elements may contain deleted ones?

Reply:
I didn't got you, did you mean it's better to handle `removed` (L215) before `added` (L211)?

Either is ok.

Reply:
Yes, what I mean is to delete first and then add, so as to prevent the elements from being cleared when deleting and adding elements that contain the same content.

Reply:
done